lib/flutter/flutter_launch.dart

```diff
@@ -148,78 +156,54 @@ class _LaunchInstance {
-            'localhost',
-            _observatoryPort,
-            uriTranslator: translator
-          ).then((_) {
-            _launch.servicePort.value = _observatoryPort;
-          }).catchError((e) {
-            _launch.pipeStdio(
-              'Unable to connect to the Observatory at port ${_observatoryPort}.\n',
-              error: true
-            );
-          });
-        });
-      }
-
-      // Chain 'flutter logs'.
-      List<String> logsArgs = <String>['logs'];
-
-      // Just log from the currently selected device.
-      if (_device != null) {
-        logsArgs.add('--device-id');
-        logsArgs.add(_device.id);
-      }
+    _runner.onStdout.listen((String str) {
+      _watchForObservatoryMessage(str);
+      _launch.pipeStdio(str);
+    });
+    _runner.onStderr.listen((String str) => _launch.pipeStdio(str, error: true));
+    _runner.onExit.then((code) => _launch.launchTerminated(code));
 
-      _runner = _flutter(flutter, logsArgs, cwd: project.path);
-      _runner.execStreaming();
-      _runner.onStdout.listen((str) => _launch.pipeStdio(str));
-      _runner.onStderr.listen((str) => _launch.pipeStdio(str, error: true));
+    return _launch;
+  }
 
-      // Don't return the future here.
-      _runner.onExit.then((code) => _launch.launchTerminated(code));
-    } else {
-      _launch.launchTerminated(code);
+  void _watchForObservatoryMessage(String str) {
+    // "Observatory listening on http://127.0.0.1:8100"
+    if (!str.startsWith('Observatory listening on http'))
+      return;
+
+    if (_observatoryPort != null && _launch.servicePort.value == null) {
+      new Future.delayed(new Duration(milliseconds: 100), () {
+        FlutterUriTranslator translator = new FlutterUriTranslator(_launch.project?.path);
+        ObservatoryDebugger.connect(
+          _launch,
+          'localhost',
+          _observatoryPort,
+          uriTranslator: translator
+        ).then((_) {
+          _launch.servicePort.value = _observatoryPort;
+        }).catchError((e) {
+          _launch.pipeStdio(
+            'Unable to connect to the Observatory at port ${_observatoryPort}.\n',
+            error: true
+          );
+        });
+      });
     }
-
-    return _launch;
   }
 
   Future _kill() {
     if (_runner == null) {
       _launch.launchTerminated(1);
       return new Future.value();
     } else {
-      // Run flutter stop.
-      FlutterTool flutter = _flutterSdk.sdk.flutterTool;
-
-      List<String> args = <String>['stop'];
+      // Tell the flutter run --resident process to quit.
+      _runner.write('q');
 
-      if (_device != null) {
-        args.add('--device-id');
-        args.add(_device.id);
-      }
-
-      ProcessRunner flutterStop = _flutter(flutter, args, cwd: project.path);
-      flutterStop.execSimple().catchError((e) {
-        _logger.finer('Error from flutter stop', e);
+      return new Future.delayed(new Duration(milliseconds: 250), () {
+        return _runner.kill();
```

Comment:
Will _runner ever be null here? Is there a race condition? Or is _runner never set to null once it is assigned?